dali/base/sysinfologger.cpp

```diff
@@ -0,0 +1,648 @@
+    Owned<IRemoteConnection> conn;
+    bool updateable = false;
+    Owned<IPropertyTree> root;
+    Owned<IPropertyTreeIterator> msgIter;
+    CSysInfoLoggerMsg infoMsg;
+
+    bool ensureMatch()
+    {
+        if (filter->hasDateRange())
+        {
+            for (; msgIter->isValid(); msgIter->next())
+            {
+                unsigned __int64 ts = msgIter->query().getPropInt64(ATTR_TIMESTAMP, 0);
+                if (filter->isInDateRange(ts))
+                    return true;
+            }
+        }
+        return msgIter->isValid();
+    }
+
+public:
+    CSysInfoLoggerMsgIterator(ISysInfoLoggerMsgFilter * _filter, bool _updateable=false) : filter(_filter), updateable(_updateable)
+    {
+    }
+    CSysInfoLoggerMsg & queryInfoLoggerMsg()
+    {
+        return infoMsg.set(&(msgIter->get()), root.get(), updateable);
+    }
+    virtual ISysInfoLoggerMsg & query() override
+    {
+        return queryInfoLoggerMsg();
+    }
+    virtual bool first() override
+    {
+        unsigned mode = RTM_LOCK_READ | (updateable?RTM_LOCK_WRITE : 0);
+        conn.setown(querySDS().connect(SYS_INFO_ROOT, myProcessSession(), mode, SDS_LOCK_TIMEOUT));
+        if (!conn)
+            return false;
+        root.setown(conn->getRoot());
+
+        StringBuffer xpath;
+        filter->getQualifierXPathFilter(xpath);
+
+        msgIter.setown(root->getElements(xpath.str()));
+        if (!msgIter->first())
+            return false;
+        return ensureMatch();
+    }
+    virtual bool next() override
+    {
+        if (msgIter->next())
+            return false;
+        return ensureMatch();
+    }
+    virtual bool isValid() override
+    {
+        return msgIter ? msgIter->isValid() : false;
+    }
+};
+
+ISysInfoLoggerMsgIterator * createSysInfoLoggerMsgIterator(bool visibleOnly, bool hiddenOnly, unsigned year, unsigned month, unsigned day)
+{
+    Owned<CSysInfoLoggerMsgFilter> filter = new CSysInfoLoggerMsgFilter(visibleOnly, hiddenOnly, year, month, day);
+    return new CSysInfoLoggerMsgIterator(filter.getClear(), false);
+}
+
+ISysInfoLoggerMsgIterator * createSysInfoLoggerMsgIterator(ISysInfoLoggerMsgFilter * msgFilter)
+{
+    return new CSysInfoLoggerMsgIterator(msgFilter, false);
+}
+
+// returns messageId
+unsigned __int64 logSysInfoError(const LogMsgCategory & cat, LogMsgCode code, const char *source, const char * msg, unsigned __int64 ts)
+{
+    if (ts==0)
+        ts = getTimeStampNowValue();
+
+    if (isEmptyString(source))
+        source = "unknown";
+    Owned<IRemoteConnection> conn = querySDS().connect(SYS_INFO_ROOT, myProcessSession(), RTM_LOCK_WRITE|RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT);
+    if (!conn)
+        throw makeStringExceptionV(-1, "logSysInfoLogger: unable to create connection to '%s'", SYS_INFO_ROOT);
+
+    StringBuffer xpath;
+    unsigned year, month, day;
+    extractDate(ts, year, month, day);
+    xpath.appendf("m%04u%02u/d%02u[" ATTR_VERSION "='%s']", year, month, day, SYS_INFO_VERSION);
```

Comment:
I have still have a nagging concern that this is going to cause some SDS pain/inefficiency.
It will mean that each new message, will cause every /SysLogs/mYYYYMM to be downloaded, and every dDD unter the month.

Now, if there aren't a huge amount, or not that frequently added, then it's not too bad, but it would still be best to avoid.

instead could create a 2nd connection that creates the final node:
```
    xpath.appendf("%s/m%04u%02u/%s/d%02u/%s", SYS_INFO_ROOT, year, month, SYS_INFO_VERSION, day, MSG_NODE);
    // NB: this will create intermediate nodes if missing
    Owned<IRemoteConnection> conn = querySDS().connect(SYS_INFO_ROOT, myProcessSession(), RTM_CREATE_ADD, SDS_LOCK_TIMEOUT);
    assertex(conn);
    IPropertryTree *dayPT = conn->queryRoot();
    // could do this way, but prob. better to have a CSysInfoLoggerMsg::setPTree(IPropertyTree *target, ....)
    Owned<IPropertyTree> msgTree = CSysInfoLoggerMsg::createMsgPTree(id, cat, code, source, msg, ts, false);
    dayPT->setPropTree(nullptr, msgTree); // replaces
    dayPT->setPropInt(ATTR_LASTID, ++id);
```

This would avoid any unnecessary layers being downloaded.
NB: it means that the day SYS_INFO_VERSION version, would have to be either it's own node, or could be part of the day node name itself.

Less concerned in the read case, because I'm assuming that the normal case is to walk over multiple messages anyway.

Reply:
Seeking clarification: does querySDS().connect or conn->queryRoot(), download that entire root path? Or is the path only downloaded when conn->queryRoot()->queryPropTree() is called?

The @lastId used is stored with the day node.   Would the querying of the @lastId attribute day pt download the entire day PT? 


@jakesmith

Reply:
>Seeking clarification: does querySDS().connect or conn->queryRoot(), download that entire root path? 

no, there's a lazy fetch mechanism at each level.
Neither the connection nor queryPropTree will cause layers to be fetch at that point, but any interaction with children of a ptree node, will cause all the children of that node to be downloaded, including if a new node is added to a node.

>The @lastid used is stored with the day node. Would the querying of the @lastid attribute day pt download the entire day PT?

Yes, tt would download all child nodes (and attributes).

One issue with the suggested code above is that you won't have access to the <day>/@lastId
Since it has a write lock to SYS_INFO_ROOT, it could get/update a global lastId there (you have plenty of bits available in the logid size, for a 2^32 to be combined in makeMessageId).